modules/network/src/main/java/org/apache/ignite/internal/network/netty/NettyServer.java

```diff
@@ -260,6 +254,29 @@ public NettyServer(
         }
     }
 
+    /**
+     * Try bind this server to the port.
+     *
+     * @param port Target port.
+     * @param endPort Last port that server can be bound to.
+     * @param fut Future.
+     */
+    private void tryBind(int port, int endPort, CompletableFuture<Channel> fut) {
+        if (port > endPort)
+            fut.completeExceptionally(new IllegalStateException("No available port in range"));
+
+        NettyUtils.toChannelCompletableFuture(bootstrap.bind(port)).handle((channel, throwable) -> {
```

Comment:
I would suggest using the `bootstrap.bind(port).addListener` method instead, because otherwise you are creating rather pointless intermediate Future instances.

Reply:
also, `whenComplete` is more suitable here